### PR TITLE
Implement dashboard app

### PR DESCRIPTION
**Pull request by zolga**
The app was tested manually in local and after deploy.
[live link to the app](https://666433291-05d54161-592d-408a-9bfe-6ba15b3ee55c.shiftjx.com/)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,68 +1,11 @@
-This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
+[![CircleCI](https://circleci.com/gh/shiftjs/dashboard.svg?style=svg)](https://circleci.com/gh/shiftjs/dashboard)
```
Comment (bergundy) [inline README.md]: This URL is probably broken now since we're under `reshufflehq` and not `shiftjs`.

File: README.md [hunk 2]
```diff
@@ -1,68 +1,11 @@
...
 
-## Available Scripts
+Keywords: dashboard, classes, shiftjs, react-bootstrap, fullstack, promise, async, chartjs
```
Comment (bergundy) [inline README.md]: Let's just replace all instances of shiftjs to reshuffle please.

File: package.json [hunk 3]
```diff
@@ -1,24 +1,57 @@
...
+    "react-scripts": "3.0.1",
+    "react-test-renderer": "^16.9.0",
+    "react-transition-group": "^4.2.1",
+    "shift-cli": "^0.2.0",
+    "typescript": "^3.5.3"
   },
   "scripts": {
-    "start": "cross-env EXTEND_ESLINT=true react-scripts start",
-    "build": "cross-env EXTEND_ESLINT=true react-scripts build",
+    "start": "react-scripts start",
```
Comment (bergundy) [inline package.json]: Why was this changed?

File: public/manifest.json [hunk 4]
```diff
@@ -1,6 +1,10 @@
 {
   "short_name": "React App",
-  "name": "Create React App Sample",
+  "name": "Counter Application Using React Class Components",
```
Comment (bergundy) [inline public/manifest.json]: The short name needs to be changed too.

File: src/containers/Dashboard.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
+import React from 'react';
+import { Chart } from 'primereact/chart';
+import '@binaris/shift-babel-macro/macro';
```
Comment (bergundy) [inline src/containers/Dashboard.js]: Can you put this is the first import please?

File: backend/dashboard-server.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+/* @expose */
+export async function setStorageData(data) {
+  return await update('chartData', (prevData = []) => data);
+}
+
+/*
+  Fetch chart data from db
+ */
+/* @expose */
+export async function getStorageData() {
```
Comment (bergundy) [inline backend/dashboard-server.js]:
```suggestion
export async function getStoredData() {
```

File: backend/dashboard-server.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    Math.floor(Math.random() * 100 + 1),
+    Math.floor(Math.random() * 100 + 1),
+  ];
+}
+
+/*
+  Save chart data in db
+ */
+/* @expose */
+export async function setStorageData(data) {
```
Comment (bergundy) [inline backend/dashboard-server.js]:
```suggestion
export async function initData(data) {
```

File: backend/dashboard-server.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  return await update('chartData', (prevData = []) => data);
```
Comment (bergundy) [inline backend/dashboard-server.js]: In this use case it might be better to just create the initial data without the option of updating.
```suggestion
  return await create('chartData', data);
```

Comment (bergundy) [reply backend/dashboard-server.js]: Also add the missing import

File: backend/dashboard-server.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+/*
+  Fetch chart data from db
+ */
+/* @expose */
+export async function getStorageData() {
+  return await get('chartData');
```
Comment (bergundy) [inline backend/dashboard-server.js]: No need to `await` if you're returning the value.

File: backend/dashboard-server.js (hunk 8 above)
Comment (bergundy) [reply backend/dashboard-server.js]: No need to await if you're returning the value.

File: src/containers/Dashboard.js [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+import Button from 'react-bootstrap/Button';
+
+import {
+  generateRemoteData,
+  setStorageData,
+  getStorageData,
+} from '../../backend/dashboard-server';
+import { Display } from '../constants/enums';
+
+export default class Dashboard extends React.Component {
```
Comment (bergundy) [inline src/containers/Dashboard.js:19]: Can we please use functional components with hooks?
We generally prefer this form.

File: src/containers/Dashboard.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      chart_data: {
```
Comment (bergundy) [inline src/containers/Dashboard.js]: JS keys are usually in camel case.

File: src/containers/Dashboard.js [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+      display: Display.FRONTEND,
+    };
+    this.generateLocalData = this.generateLocalData.bind(this);
+    this.getRemoteData = this.getRemoteData.bind(this);
+    this.fetchFromStorageData = this.fetchFromStorageData.bind(this);
+    this.initStorageData();
+  }
+
+  initStorageData() {
+    setStorageData(this.state.chart_data.datasets[0].data);
```
Comment (bergundy) [inline src/containers/Dashboard.js]: Can we initialize to something different from the local data so it's clear that something changed when fetching from remote?

File: src/containers/Dashboard.js [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+            hoverBackgroundColor: ['#FF6384', '#13bdd8', '#FFCE56'],
+          },
+        ],
+      },
+      display: Display.FRONTEND,
+    };
+    this.generateLocalData = this.generateLocalData.bind(this);
+    this.getRemoteData = this.getRemoteData.bind(this);
+    this.fetchFromStorageData = this.fetchFromStorageData.bind(this);
+    this.initStorageData();
```
Comment (bergundy) [inline src/containers/Dashboard.js]: You're not supposed to have side-effects here, use `componentDidMount` instead or better yet `useEffect` hook when rewriting to a functional component.

File: src/containers/Dashboard.js (hunk 12 above)
Comment (bergundy) [reply src/containers/Dashboard.js]: We ignore the possibility of an error here, we should catch the error and report it to the user, e.g: https://github.com/binaris/shiftjs/blob/262a886958214313f1bbed815d843b1513c6cb18/app-testsuite/app/src/App.js#L5

File: src/containers/Dashboard.js [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+  /* Generate random chart data from backend*/
+  async getRemoteData() {
+    const newdata = Object.assign({}, this.state.chart_data);
+    newdata.datasets[0].data = await generateRemoteData();
+    this.setState({ display: Display.BACKEND });
+    this.setState({ chart_data: newdata });
+  }
+
+  /* Fetch random chart data from db*/
+  async fetchFromStorageData() {
```
Comment (bergundy) [inline src/containers/Dashboard.js]:
```suggestion
  async fetchStoredData() {
```

File: src/containers/Dashboard.js [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+      Math.floor(Math.random() * 100 + 1),
+      Math.floor(Math.random() * 100 + 1),
+    ];
+    this.setState({ display: Display.FRONTEND });
+    this.setState({ chart_data: newdata });
+  }
+
+  /* Generate random chart data from backend*/
+  async getRemoteData() {
+    const newdata = Object.assign({}, this.state.chart_data);
```
Comment (bergundy) [inline src/containers/Dashboard.js]: This is a shallow copy, please use something like [immer.produce](https://github.com/immerjs/immer#reactsetstate-example) in order to create a new value.
Same goes for the other state update functions.

File: src/containers/Dashboard.js [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    newdata.datasets[0].data = await generateRemoteData();
```
Comment (bergundy) [inline src/containers/Dashboard.js]: No error handling here will result in an UnhandledRejection.

File: src/containers/Dashboard.js [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    this.setState({ display: Display.BACKEND });
+    this.setState({ chart_data: newdata });
+  }
+
+  /* Fetch random chart data from db*/
+  async fetchFromStorageData() {
+    const newdata = Object.assign({}, this.state.chart_data);
+    newdata.datasets[0].data = await getStorageData();
```
Comment (bergundy) [inline src/containers/Dashboard.js]: Again no error handling.

File: src/containers/Dashboard.js (hunk 16 above)
Comment (bergundy) [reply src/containers/Dashboard.js]: Also it would be nice to have a loading state even though we expect the backend to be fast.

File: package.json [hunk 18]
```diff
@@ -1,24 +1,57 @@
...
+    "bootstrap": "^4.3.1",
+    "chart.js": "^2.8.0",
+    "classnames": "^2.2.6",
+    "primeicons": "^1.0.0",
+    "primereact": "^3.1.7",
+    "react": "^16.9.0",
+    "react-bootstrap": "^1.0.0-beta.12",
+    "react-dom": "^16.9.0",
+    "react-scripts": "3.0.1",
+    "react-test-renderer": "^16.9.0",
```
Comment (bergundy) [inline package.json]: I don't see this used in the code base

File: package.json [hunk 19]
```diff
@@ -1,24 +1,57 @@
...
+    "react-transition-group": "^4.2.1",
```
Comment (bergundy) [inline package.json]: Or this

File: package.json [hunk 20]
```diff
@@ -1,24 +1,57 @@
...
+    "shift-cli": "^0.2.0",
+    "typescript": "^3.5.3"
```
Comment (bergundy) [inline package.json]: Or this

File: package.json [hunk 21]
```diff
@@ -1,24 +1,57 @@
...
-    "react": "^16.8.6",
-    "react-dom": "^16.8.6",
-    "react-scripts": "3.1.1"
+    "@binaris/shift-leveldb-server": "0.0.2",
+    "@binaris/shift-local-proxy": "0.2.0",
+    "@binaris/shift-server-function": "0.0.2",
+    "bootstrap": "^4.3.1",
+    "chart.js": "^2.8.0",
+    "classnames": "^2.2.6",
+    "primeicons": "^1.0.0",
```
Comment (bergundy) [inline package.json]: I don't see this used in the code base

File: package.json (hunk 19 above)
Comment (bergundy) [reply package.json]: Please verify all of the dependencies listed here are actually dependencies.

**Review by bergundy — CHANGES_REQUESTED** [review]
Thanks!
The overall direction is good and the app looks good.
There are some subtle things that were missed regarding unhandled promise rejections, lack of loading and error states and mutating state.

File: src/containers/Dashboard.js (hunk 5 above)
Comment (michaeladda) [reply src/containers/Dashboard.js]: also, the new macro is `@binaris/shift-code-transform/macro` not `shift-babel-macro`.
you can see the always up-to-date test application we're using at https://github.com/binaris/shiftjs/tree/master/app-testsuite/app
(and its frontend https://github.com/binaris/shiftjs/blob/master/app-testsuite/app/src/App.js)

File: public/manifest.json (hunk 4 above)
Comment (michaeladda) [reply public/manifest.json]: Counter -> Dashboard

File: package.json [hunk 22]
```diff
@@ -1,24 +1,57 @@
...
   },
   "scripts": {
-    "start": "cross-env EXTEND_ESLINT=true react-scripts start",
-    "build": "cross-env EXTEND_ESLINT=true react-scripts build",
+    "start": "react-scripts start",
+    "build": "react-scripts build",
     "test": "react-scripts test",
     "eject": "react-scripts eject",
-    "lint": "eslint backend && stylelint src/**/*.css"
+    "lint": "eslint backend --fix && eslint src/* --fix"
```
Comment (michaeladda) [inline package.json]:
```suggestion
    "lint": "eslint backend --fix && eslint src --fix"
```

File: package.json [hunk 23]
```diff
@@ -1,24 +1,57 @@
...
     "@binaris/shift-code-transform": "0.0.2",
-    "@binaris/shift-db": "^0.3.1",
+    "@binaris/shift-db": "0.5.0",
     "@binaris/shift-fetch-runtime": "0.0.4",
-    "@binaris/shift-local-proxy": "0.0.11",
-    "@binaris/shift-server-function": "0.0.1",
-    "react": "^16.8.6",
-    "react-dom": "^16.8.6",
-    "react-scripts": "3.1.1"
+    "@binaris/shift-leveldb-server": "0.0.2",
```
Comment (michaeladda) [inline package.json]: `shift-leveldb-server` is no longer required (it is imported via `shift-local-proxy`)

**Review by michaeladda — CHANGES_REQUESTED** [review]
Thanks!
The dashboard is visually stunning!

Requested changes are wrong names (counter, shift), obsolete package, and react side-effects in the constructor (see https://reactjs.org/docs/react-component.html#constructor specifically
`Avoid introducing any side-effects or subscriptions in the constructor. For those use cases, use componentDidMount() instead.`)

While the team prefers functional components with hooks, classes are acceptable too, and componentDidMount is fine to solve the side effects issue. This is a suggestion and not a request.

File: README.md (hunk 1 above)
Comment (zolga) [reply README.md]: changed, thanks.

File: backend/dashboard-server.js (hunk 8 above)
Comment (zolga) [reply backend/dashboard-server.js]: Done, Thanks.

File: backend/dashboard-server.js (hunk 6 above)
Comment (zolga) [reply backend/dashboard-server.js]: Done, Thanks.

File: backend/dashboard-server.js (hunk 9 above)
Comment (zolga) [reply backend/dashboard-server.js]: removed, thanks.

File: package.json (hunk 23 above)
Comment (zolga) [reply package.json]: Removed, thanks!

File: package.json (hunk 21 above)
Comment (zolga) [reply package.json]: remove. thanks.

File: package.json (hunk 18 above)
Comment (zolga) [reply package.json]: I needed this to pass unit-tests: [discussion](https://binarisio.slack.com/archives/C4VEBR3E3/p1568053476052200)

File: package.json (hunk 19 above)
Comment (zolga) [reply package.json]: Done, I used  `npx depcheck` and removed unused dependencies. Thanks!

File: package.json (hunk 20 above)
Comment (zolga) [reply package.json]: changed , related to my comment above.

File: package.json (hunk 3 above)
Comment (zolga) [reply package.json]: Changed back, probably bad merge.

File: package.json (hunk 22 above)
Comment (zolga) [reply package.json]: fixed. thanks!

File: public/manifest.json (hunk 4 above)
Comment (zolga) [reply public/manifest.json]: Changed, thanks.

File: src/containers/Dashboard.js (hunk 5 above)
Comment (zolga) [reply src/containers/Dashboard.js]: > Can you put this is the first import please?

@bergundy  From what I tested it's working well if it's not in the top, can you explain the logic behind your request, please?

Comment (zolga) [reply src/containers/Dashboard.js]: > also, the new macro is `@binaris/shift-code-transform/macro` not `shift-babel-macro`.
> you can see the always up-to-date test application we're using at https://github.com/binaris/shiftjs/tree/master/app-testsuite/app
> (and its frontend https://github.com/binaris/shiftjs/blob/master/app-testsuite/app/src/App.js)
Changed, Thanks!

File: src/containers/Dashboard.js (hunk 11 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Done. thanks!

File: src/containers/Dashboard.js (hunk 14 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Renamed, Thanks.

File: backend/dashboard-server.js (hunk 7 above)
Comment (zolga) [reply backend/dashboard-server.js]: Cool, changed.

File: src/containers/Dashboard.js (hunk 13 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Totally agree, but with this app, I will keep the class approach so we have more then one type of react app.
I moved the call to componentDidMount
Thanks!

File: src/containers/Dashboard.js (hunk 16 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Added error handling + loading message & spinner. thanks!

File: src/containers/Dashboard.js (hunk 17 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Done, Thanks.

**Comment by zolga** [comment]
Thank you for your feedback!
The code now is much cleaner and readable - thanks!

I tried to manage all the commits by each feedback / CR - Please take another look @bergundy  @michaeladda

File: src/containers/Dashboard.js [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+      },
+      display: Display.FRONTEND,
+    };
+    this.generateLocalData = this.generateLocalData.bind(this);
+    this.getRemoteData = this.getRemoteData.bind(this);
+    this.fetchStoredData = this.fetchStoredData.bind(this);
+  }
+
+  componentDidMount() {
+    initData(this.state.chartData.datasets[0].data);
```
Comment (bergundy) [inline src/containers/Dashboard.js]: Please handle the potential error here.
You might want to (not required) to start with loading state and only transition to frontend state after `initData` is complete similarly to `getRemoteData`.

File: src/containers/Dashboard.js [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+          <Spinner animation='border' variant='info' className='mr-2' />
+          <span className='pl-1'>Loading...</span>
+        </Container>
+      </Container>
+    );
+  }
+
+  /* Generate random chart data */
+  generateLocalData() {
+    const newdata = Object.assign({}, this.state.chartData);
```
Comment (bergundy) [inline src/containers/Dashboard.js]: Again, this doesn't perform a deep copy of the object.
Try this in your browser console:
```js
const chartData = { datasets: [ { data: [1,2,3] } ] };
const newdata = Object.assign({}, chartData);
newdata.datasets[0].data = [3,4,5];
chartData.datasets[0].data;
```

You'll see that the original chartData was modified.

I suggest one of these versions:
```js
import produce from 'immer';

const newData = produce(this.state.chartData, draftState => {
  draftState.datasets[0].data = [/* ... */];
})
```
Or
```js
const newData = {
  ...this.state.chartData,
  datasets: [
    {
      ...this.state.chartData.datasets[0],
      data: [/* ... */],
    },
  ],
};
```
Which is the equivalent of
```js
const newData = Object.assign({}, this.state.chartData, {
  datasets: [
    Object.assign({}, this.state.chartData.datasets[0], { data: [/* ... */] }),
  ],
};
```

File: src/containers/Dashboard.js [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+    newdata.datasets[0].data = [
+      Math.floor(Math.random() * 100 + 1),
+      Math.floor(Math.random() * 100 + 1),
+      Math.floor(Math.random() * 100 + 1),
+    ];
+    this.setState({ display: Display.FRONTEND });
+    this.setState({ chartData: newdata });
+  }
+
+  /* Generate random chart data from backend*/
```
Comment (bergundy) [inline src/containers/Dashboard.js]:
```suggestion
  /** Generate random chart data from backend */
```
To make it a docstring.

File: src/containers/Dashboard.js [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+    }
+  }
+
+  /* Fetch random chart data from db*/
+  async fetchStoredData() {
+    try {
+      this.setState({ display: Display.LOADING });
+      const newdata = Object.assign({}, this.state.chartData);
+      newdata.datasets[0].data = await getStoredData();
+      this.setState({ display: Display.DB });
```
Comment (bergundy) [inline src/containers/Dashboard.js]: These could have been unified into a single call to `setState`

File: src/containers/Dashboard.js (hunk 26 above)
Comment (bergundy) [reply src/containers/Dashboard.js]: Same applies to the comments above.

**Review by bergundy — COMMENTED** [review]
Thanks!
I see most of the comments were addressed.
I think we can merge after another round of fixes.
Well done.

File: src/containers/Dashboard.js (hunk 24 above)
Comment (zolga) [reply src/containers/Dashboard.js]: I don't think we should have a loader at this point, and I will explain: 
This function basically creates the document in the DB, we are showing this data only when the button was pressed (and then we are displaying the spinner/error respectively).
There is nothing that is loading and we are not waiting for something to happen.

I added try..catch to this block of code, in case our call will get Promise rejection. Currently, we won't support messaging/logs to show the user, so I ended up adding log to the console in case there is an error.

File: src/containers/Dashboard.js (hunk 25 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Sorry, I missed this comment in your prev review. 
Thanks for the examples! I decided to keep it simple and not use additional npm package. 
I fixed by using` ...this.state.chartData` .

File: src/containers/Dashboard.js (hunk 26 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Fixed in all files, thanks.

File: src/containers/Dashboard.js (hunk 27 above)
Comment (zolga) [reply src/containers/Dashboard.js]: Agree, Fixed!

**Review by zolga — COMMENTED** [review]
Thanks for the great feedback! 
All the comment where fixed.
I missed your comment from your previews review about deep cloning the object -  sorry! fixed it in this push. 

The app was tested manually in the browser.
@bergundy @michaeladda

File: src/containers/Dashboard.js [hunk 28]
```diff
@@ -39,7 +39,11 @@ export default class Dashboard extends React.Component {
   }
 
   componentDidMount() {
-    initData(this.state.chartData.datasets[0].data);
+    try {
+      initData(this.state.chartData.datasets[0].data);
+    } catch (err) {
+      console.log('Error initializing data to DB');
```
Comment (bergundy) [inline src/containers/Dashboard.js:45]: Perhaps use `console.error` for this.

**Review by bergundy — APPROVED** [review]
Thanks!

**Review by michaeladda — APPROVED** [review]
Super!
Thanks!

BTW, please notice the app url changed to https://666433291-05d54161-592d-408a-9bfe-6ba15b3ee55c.reshuffle.app/ as we're shifting to reshuffle :-p

**Comment by zolga** [comment]
Thanks!
